File: backend/services/scraper/zepto_live.py
import asyncio
import sys
from playwright.async_api import async_playwright
from config.settings import MAX_PRODUCTS

async def scrape_zepto_live(product_name):

    
    url = f"https://www.zeptonow.com/search?query={product_name}"

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True, args=["--disable-blink-features=AutomationControlled"])
        context = await browser.new_context(
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        )
        page = await context.new_page()

        try:
            await page.goto(url, timeout=20000)

            await page.wait_for_selector('a[href*="/pn/"]', timeout=10000)

            product_candidates = await page.query_selector_all('a[href*="/pn/"]')

            results = []

            import re
            for product in product_candidates[:20]:
                try:
                    name_el = await product.query_selector('[data-slot-id="ProductName"]')
                    name = await name_el.inner_text() if name_el else None
                    if not name:
                        continue

                    # Price is usually the first span in EdlpPrice
                    price_el = await product.query_selector('[data-slot-id="EdlpPrice"] span:nth-of-type(1)')
                    price = await price_el.inner_text() if price_el else None
                    
                    # MRP is the second span
                    mrp_el = await product.query_selector('[data-slot-id="EdlpPrice"] span:nth-of-type(2)')
                    mrp = await mrp_el.inner_text() if mrp_el else None

                    # Discount parsing
                    discount = "0% OFF"
                    all_text = await product.inner_text()
                    match = re.search(r"(\d+%\s*OFF)", all_text, re.IGNORECASE)
                    if match:
                        discount = match.group(1)
                    
                    if price:
                        price = price.replace("₹", "").strip()
                    if mrp:
                        mrp = mrp.replace("₹", "").strip()
                    if "OFF" in discount:
                        discount = discount.upper().replace("OFF", "").strip()

                    # Extract real product image
                    img_el = await product.query_selector("img")
                    product_image = await img_el.get_attribute("src") if img_el else None
                    if not product_image or product_image.startswith("data:"):
                        product_image = await img_el.get_attribute("data-src") if img_el else None
                    product_image = product_image or "https://via.placeholder.com/150"

                    results.append({
                        "platform": "zepto",
                        "product_name": name,
                        "description": name,
                        "weight": "N/A",
                        "price": price,
                        "mrp": mrp,
                        "discount": discount,
                        "store_name": "Zepto",
                        "product_image": product_image
                    })
                except Exception:
                    continue

            await browser.close()
            return results

        except Exception as e:
            import traceback
            # Capture HTML for debugging
            try:
                content = await page.content()
                with open("zepto_debug.html", "w", encoding="utf-8") as f:
                    f.write(content)
            except:
                pass

            with open("scraper_error.log", "a") as f:
                f.write(f"Zepto Error for {product_name}:\n")
                f.write(traceback.format_exc())
                f.write("\n" + "="*50 + "\n")
            await browser.close()
            return []

File: node_backend/services/scraper/zeptoLive.js
/**
 * Zepto Live Scraper — equivalent to services/scraper/zepto_live.py
 * Quick scrape of Zepto search results using Playwright.
 */
const { MAX_PRODUCTS } = require('../../config/settings');

async function scrapeZeptoLive(productName) {
  const { chromium } = require('playwright');
  const url = `https://www.zeptonow.com/search?query=${encodeURIComponent(productName)}`;

  const browser = await chromium.launch({
    headless: true,
    args: ['--disable-blink-features=AutomationControlled'],
  });
  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
  });
  const page = await context.newPage();

  try {
    await page.goto(url, { timeout: 20000 });
    await page.waitForSelector('a[href*="/pn/"]', { timeout: 10000 });

    const productCandidates = await page.$$('a[href*="/pn/"]');
    const results = [];

    for (const product of productCandidates.slice(0, 20)) {
      try {
        const nameEl = await product.$('[data-slot-id="ProductName"]');
        const name = nameEl ? await nameEl.innerText() : null;
        if (!name) continue;

        const priceEl = await product.$('[data-slot-id="EdlpPrice"] span:nth-of-type(1)');
        let price = priceEl ? await priceEl.innerText() : null;

        const mrpEl = await product.$('[data-slot-id="EdlpPrice"] span:nth-of-type(2)');
        let mrp = mrpEl ? await mrpEl.innerText() : null;

        // Discount parsing
        let discount = '0% OFF';
        const allText = await product.innerText();
        const match = allText.match(/(\d+%\s*OFF)/i);
        if (match) discount = match[1];

        if (price) price = price.replace('₹', '').trim();
        if (mrp) mrp = mrp.replace('₹', '').trim();
        if (discount.includes('OFF')) discount = discount.toUpperCase().replace('OFF', '').trim();

        // Extract product image
        const imgEl = await product.$('img');
        let productImage = imgEl ? await imgEl.getAttribute('src') : null;
        if (!productImage || productImage.startsWith('data:')) {
          productImage = imgEl ? await imgEl.getAttribute('data-src') : null;
        }
        productImage = productImage || 'https://via.placeholder.com/150';

        // Zepto product cards are <a href="/pn/..."> elements
        const href = await product.getAttribute('href') || '';
        const productUrl = href.startsWith('/') ? `https://www.zeptonow.com${href}` : href;

        results.push({
          platform: 'zepto',
          product_name: name,
          description: name,
          weight: 'N/A',
          price,
          mrp,
          discount,
          store_name: 'Zepto',
          product_image: productImage,
          product_url: productUrl || '',
        });
      } catch {
        continue;
      }
    }

    // Limit to MAX_PRODUCTS
    await browser.close();
    return results.slice(0, MAX_PRODUCTS);
  } catch (e) {
    console.log(`    [Zepto] Scraping error: ${e.message}`);
    try { await browser.close(); } catch { /* ignore */ }
    return [];
  }
}

module.exports = { scrapeZeptoLive };
